MoreLinq/ExceptAll.cs

```diff
@@ -0,0 +1,186 @@
+        /// whose keys are not in the second sequence, according to a given key selector.
+        /// </summary>
+        /// <typeparam name="TSource">The type of source and result elements.</typeparam>
+        /// <typeparam name="TKey">The type of the key.</typeparam>
+        /// <param name="first">The sequence of potentially included elements.</param>
+        /// <param name="second">The sequence of keys which may prevent elements in
+        /// <paramref name="first"/> from being returned.</param>
+        /// <param name="keySelector">The mapping from source element to key.</param>
+        /// <returns>
+        /// Sequence of elements from first sequence,
+        /// whose keys are not in second sequence.
+        /// </returns>  
+        /// <remarks>
+        /// This is a sequence operation; if multiple elements in <paramref name="first"/> have
+        /// equal keys, all such elements are returned.
+        /// This operator uses deferred execution and streams the results, although
+        /// a set of keys from <paramref name="second"/> is immediately selected and retained.
+        /// </remarks>
+        public static IEnumerable<TSource> ExceptAllKeys<TSource, TKey>(this IEnumerable<TSource> first,
+            IEnumerable<TKey> second,
+            Func<TSource, TKey> keySelector) {
+
+            if (first == null) throw new ArgumentNullException("first");
+            if (second == null) throw new ArgumentNullException("second");
+            if (keySelector == null) throw new ArgumentNullException("keySelector");
+
+            return ExceptAllKeysImpl(first, second, keySelector, null);
+        }
+
+        /// <summary>
+        /// Returns the sequence of elements from the first collection,
+        /// whose keys are not in the second collection, according to a given key selector.
+        /// </summary>
+        /// <typeparam name="TSource">The type of source and result elements.</typeparam>
+        /// <typeparam name="TKey">The type of the key.</typeparam>
+        /// <param name="first">The sequence of potentially included elements.</param>
+        /// <param name="second">The sequence of keys which may prevent elements in
+        /// <paramref name="first"/> from being returned.</param>
+        /// <param name="keySelector">The mapping from source element to key.</param>
+        /// <param name="keyComparer">The key comparer. If <c>null</c>, uses the default TKey equality comparer.</param>
+        /// <returns>
+        /// Sequence of elements from first sequence,
+        /// whose keys are not in second sequence.
+        /// </returns>
+        /// <remarks>
+        /// This is a sequence operation; if multiple elements in <paramref name="first"/> have
+        /// equal keys, all such elements are returned.
+        /// This operator uses deferred execution and streams the results, although
+        /// a set of keys from <paramref name="second"/> is immediately selected and retained.
```

Comment:
This is not true. The `IEnumerable` generated by the compiler does not execute anything until the first `MoveNext` call (which is why you have to do the null checks in the wrapper functions instead of the Impl one).